domain-models-runtime/src/main/java/org/folio/rest/persist/cache/CachedConnectionManager.java

```diff
@@ -109,19 +112,74 @@ private void limitCacheSize() {
 
   private Future<PgConnection> createConnectionSession(Vertx vertx, Pool pool, String schemaName, String tenantId) {
     connectionCache.setPoolSizeMetric(pool.size());
-    return pool.getConnection().compose(sqlConnection -> setRoleAndSchema(vertx, schemaName, tenantId, sqlConnection));
+    return pool
+        .getConnection()
+        .compose(sqlConnection -> setRoleAndSchema(vertx, schemaName, tenantId, sqlConnection));
   }
 
   private Future<PgConnection> setRoleAndSchema(Vertx vertx,
                                                 String schemaName,
                                                 String tenantId,
                                                 SqlConnection sqlConnection) {
-    connectionCache.incrementActive();
-    String sql = PostgresClient.DEFAULT_SCHEMA.equals(tenantId)
+    // Only wrap raw PgConnection, never wrap CachedPgConnection
+    if (sqlConnection instanceof CachedPgConnection) {
+      throw new IllegalStateException("Attempted to wrap CachedPgConnection in another CachedPgConnection");
+    }
+
+    String sql = getSetRoleAndSchemaSql(schemaName, tenantId);
+    return sqlConnection.query(sql).execute()
+        .map(r -> {
+          var cachedConnection = new CachedPgConnection(tenantId, schemaName, (PgConnection) sqlConnection,
+              this, vertx, CONNECTION_RELEASE_DELAY_SECONDS);
+
+          connectionCache.incrementActive();
+          connectionCache.incrementNewConnections();
+
+          LOG.debug("Created new connection session for tenant {}: {}", tenantId, cachedConnection.getSessionId());
+
+          return (PgConnection) cachedConnection;
+        })
+        .recover(e -> {
+          //  Clean up on failure
+          connectionCache.decrementActive();
```

Comment:
if `incrementActive()` fails or something else fails earlier, this line would decrement `active` causing a negative count right?